### ISSUE TITLE
detecting dependencies on the presence of OS packages

### ISSUE THREAD

**Issue by DanGrayson** [Macaulay2/M2#1194]
Currently we have code included in the makefiles that build the distributions (rpm, deb, dmg) for detecting dependence of M2 on shared libraries.  For the Mac, we include the libraries (or used to, and can again).  For deb and rpm, we discover the name of the appropriate package and list it as a dependency in the appropriate header file of the distribution.

But we do such thing for programs.  For example, if gfan is found on the system and we don't build it, then we currently fail to record the dependency anywhere, and we should.

Most developers don't make binary distributions, satisfying themselves with source distributions.  We used to do that long ago.  Soon, due to the efforts of Doug Torrance (@d-torrance ), we have a debian distribution, and he will take care of recording those dependencies for us.  Once we get into Homebrew (a volunteer is needed), that will also be taken care of for us.  The sooner we can stop making binary distributions, the better.

Meanwhile, what should we do?  For the mac, we could wait until gfan is to be called upon, and, if it's not there, ask the user to install it with homebrew.  That will give users practice with installing homebrew, where I hope M2 will ultimately reside, anyway.  For the deb and rpm files we make, we should continue to record the dependencies.

@mahrud - does cmake offer a canned facility for determining dependencies?

**Comment by mahrud** [comment]
If I understand the question correctly, then yes, the [CPack](https://gitlab.kitware.com/cmake/community/-/wikis/doc/cpack/PackageGenerators) utility does exactly that. Right now you can build a .dep or .rpm package with `cpack -G RPM` or `DEB` on the CMake branch. The `cmake/packaging.cmake` scripts defines the relevant variables. I haven't put much work into it, since I assumed you would want to use the configure/make system for distributions, but if you're interested in using cmake for that, I can work on a second PR for cpack.

@d-torrance have you tried building a .deb package with `cpack -G DEB`? It works well in my tests. What key results should it satisfy?

**Comment by DanGrayson** [comment]
In the long run, we don't want to be making distributions or maintaining our own binary repositories, so it's probably best not invest too much energy into that kind of thing.

Does `packaging.cmake` have the intelligence to determine which package a needed file comes from?  It seems too brief for that.

**Comment by d-torrance** [comment]
> @d-torrance have you tried building a .deb package with `cpack -G DEB`? It works well in my tests. What key results should it satisfy?

No, I haven't tried it.  I pretty much just use [pbuilder](https://manpages.debian.org/unstable/pbuilder/pbuilder.8.en.html) from [this branch](https://salsa.debian.org/science-team/macaulay2) of the git repository.  I did use the debian packages built out of `M2/distributions/deb` as a starting point, e.g., building two binary packages (the arch-independent `macaulay2-common` and the arch-dependent `macaulay2`.

A big goal for any Debian package is to be "[lintian](https://manpages.debian.org/unstable/lintian/lintian.1.en.html) clean".  Out of curiosity, I just ran it on the [Debian buster .debs for 1.15](https://faculty.math.illinois.edu/Macaulay2/Repositories/Debian/dists/buster/main/binary-amd64/index.html), and it [wasn't very clean](https://gist.github.com/d-torrance/fed3a72a910f2afacbb5897b83641064).

The draft I've been working on is pretty close to Lintian clean, except for a bunch of build paths in the documentation (#1149).

**Comment by DanGrayson** [comment]
A pull request making M2 upstream more clean in the sense of lintian would be welcome.

**Comment by d-torrance** [comment]
> A pull request making M2 upstream more clean in the sense of lintian would be welcome.

Sounds good.  A number of my recent pull requests have already taken care of a few of these (e.g., `privacy-breach-generic` and `executable-not-elf-or-script`).  I think #1144 and #1149 are the two remaining issues keeping M2 itself from being completely Lintian clean (and #1144 is an easy fix assuming the presence of `jdupes` or similar).

Most of the rest of the Lintian warnings are coming from how the .debs are put together in `M2/distributions/deb/Makefile` (e.g., `package-contains-timestamped-gzip`) or from the fact that they contain a bunch of embedded programs and libraries.

Is it worth making too many changes to `M2/distributions/deb/Makefile` if you're planning on moving away from distributing binaries anyway?

**Comment by DanGrayson** [comment]
Re: "Is it worth making too many changes to M2/distributions/deb/Makefile if you're planning on moving away from distributing binaries anyway?"

You'll probably make a clean debian distribution before we do, so I guess not.

**Comment by mahrud** [comment]
> Does `packaging.cmake` have the intelligence to determine which package a needed file comes from? It seems too brief for that.

That script only sets variables. The "intelligence" is in CPack using `dpkg-shlibdeps` for DEB (for RPM, the `rpm` tool takes care of dependencies automatically). We would only need to add other programs as suggestions or recommendations.

@d-torrance Thanks, that helps a lot! I just tried running lintian (no arguments) on the .deb package produced by CMake and after a bit of tweaking reduced it to just the following:
```
E: macaulay2: debian-changelog-file-missing
E: macaulay2: no-copyright-file
W: macaulay2: description-starts-with-leading-spaces
W: macaulay2: duplicate-font-file usr/share/Macaulay2/Visualize/fonts/glyphicons-halflings-regular.ttf also in fonts-glewlwyd
W: macaulay2: embedded-javascript-library usr/share/Macaulay2/Visualize/js/jquery-1.11.3.min.js please use libjs-jquery
W: macaulay2: embedded-javascript-library usr/share/Macaulay2/Visualize/js/underscore.js please use libjs-underscore
W: macaulay2: binary-without-manpage usr/bin/M2
W: macaulay2: binary-without-manpage usr/bin/M2-binary
W: macaulay2: script-not-executable usr/share/Macaulay2/NumericalAlgebraicGeometry/hom4ps2_in_out
```
The changelog and copyright formats are surprisingly specific. Do you add any other arguments? I'm bringing up packaging with CPack because it [helps with](https://cmake.org/cmake/help/latest/cpack_gen/deb.html#reproducible-packages) building reproducible packages, which might make things easier.

**Comment by d-torrance** [comment]
A few of those should be pretty quick fixes, e.g., there's already a Debian changelog at `M2/distributions/deb/changelog`.  CMake must not be putting it in the right spot.

The copyright files are my least favorite part of packaging, as they can get tedious (especially with a big package like Macaulay2!)  I understand the importance, though.  Here's the one I've put together:

https://salsa.debian.org/science-team/macaulay2/-/blob/master/debian/copyright

**Comment by d-torrance** [comment]
> Do you add any other arguments?

To lintian?  I generally run `lintian -I -E --pedantic` to display all of the various levels of warnings.

**Comment by mahrud** [comment]
Thanks! Do you know what should I do to get rid of the `binary-without-manpage` warning? I have `M2.1.gz` and `M2-binary.1.gz` installed in `share/man/man1`. Is there a step I'm missing?

**Comment by d-torrance** [comment]
Do you have them going into `macaulay2-common`?  Even though manpages are architecture-independent, Debian policy insists that they live in the same package as the corresponding executables.

This is how I have them split up:

```
profzoom@gloria:~/src/macaulay2/M2/debian$ cat macaulay2.install 
usr/bin
usr/lib
usr/share/man
profzoom@gloria:~/src/macaulay2/M2/debian$ cat macaulay2-common.install 
usr/share/Macaulay2
usr/share/doc
usr/share/emacs
usr/share/info/*.info
```

**Comment by mahrud** [comment]
No, I haven't split the packages.

**Comment by d-torrance** [comment]
Hrm, that's strange.  As long as they end up in `/usr/share/man/man1`, Lintian should be happy:

```
profzoom@gloria:~$ dpkg -c ~/pbuilder/sid_result/macaulay2_1.15.1.0+git42.5cc53ca+ds-1_amd64.deb | grep "\.1"
-rw-r--r-- root/root      1514 2015-09-06 00:03 ./usr/share/man/man1/M2.1.gz
lrwxrwxrwx root/root         0 2015-09-06 00:03 ./usr/share/man/man1/M2-binary.1.gz -> M2.1.gz
```

**Comment by DanGrayson** [comment]
Maybe lintian gets confused because the path starts with `./`.

**Comment by d-torrance** [comment]
This was initially fixed in https://github.com/Macaulay2/M2/commit/0d2d05966072804f66cdb9cfd040efd1b5e46204,.

That's basically out of date, though, as now the only binary distributions that we have a hand in (Homebrew, Debian/Ubuntu, and Fedora/RHEL) use hardcoded lists of dependencies in the corresponding packaging files.
